README.md
react-work-calendar
===================

This is a forked project of original [react-big-calendar](https://www.npmjs.com/package/react-big-calendar)

react-big-calendar
========================

An events calendar component built for React and made for modern browsers (read: IE10+) and uses flexbox over the classic tables-ception approach.

[__DEMO and Docs__](http://intljusticemission.github.io/react-big-calendar/examples/index.html)

Inspired by [Full Calendar](http://fullcalendar.io/).

## Use and Setup

`npm install react-big-calendar --save`

Include `react-big-calendar/lib/css/react-big-calendar.css` for styles.

## Run examples locally

* Clone this repository
* Retrieve dependencies: `npm install`
* Start: `npm run examples`
* Open [localhost:3000/examples/index.html](http://localhost:3000/examples/index.html).

### Localization and Date Formatting

`react-big-calendar` includes two options for handling the date formatting and culture localization, depending
on your preference of DateTime libraries. You can use either the [Moment.js](http://momentjs.com/) or [Globalize.js](https://github.com/jquery/globalize) localizers.

Regardless of your choice, you __must__ choose a localizer to use this library:

#### Moment.js

```js
import BigCalendar from 'react-big-calendar';
import moment from 'moment';

BigCalendar.setLocalizer(
  BigCalendar.momentLocalizer(moment)
);
```

#### Globalize.js v0.1.1

```js
import BigCalendar from 'react-big-calendar';
import globalize from 'globalize';

BigCalendar.setLocalizer(
  BigCalendar.globalizeLocalizer(globalize)
);
```


CHANGELOG.md
v0.14.12 - Wed, 06 Sep 2017 20:35:13 GMT
----------------------------------------





v0.14.11 - Wed, 06 Sep 2017 20:33:16 GMT
----------------------------------------





v0.14.7 - Mon, 22 May 2017 08:01:37 GMT
---------------------------------------





v0.14.6 - Mon, 22 May 2017 07:30:13 GMT
---------------------------------------





v0.14.5 - Mon, 22 May 2017 07:29:48 GMT
---------------------------------------





v0.14.5 - Mon, 22 May 2017 07:29:00 GMT
---------------------------------------





v0.14.4 - Mon, 22 May 2017 07:28:27 GMT
---------------------------------------





v0.14.4 - Mon, 22 May 2017 07:26:54 GMT
---------------------------------------

- [45687c9](../../commit/45687c9) [fixed] allow string names in `move()`
- [5578559](../../commit/5578559) [fixed] all day event selection



v0.14.0 - Tue, 02 May 2017 13:14:45 GMT
---------------------------------------





v0.13.0 - Wed, 22 Mar 2017 15:09:54 GMT
---------------------------------------





v0.12.3 - Sun, 15 Jan 2017 17:15:59 GMT
---------------------------------------





v0.12.2 - Sun, 15 Jan 2017 17:09:50 GMT
---------------------------------------

- [45687c9](../../commit/45687c9) [fixed] allow string names in `move()`



v0.12.1 - Thu, 12 Jan 2017 20:47:22 GMT
---------------------------------------

- [5578559](../../commit/5578559) [fixed] all day event selection



v0.12.0 - Sat, 07 Jan 2017 22:03:45 GMT
---------------------------------------





v0.11.1 - Sun, 20 Nov 2016 17:48:51 GMT
---------------------------------------





v0.11.0 - Sat, 08 Oct 2016 20:24:51 GMT
---------------------------------------





v0.10.3 - Fri, 15 Jul 2016 17:56:54 GMT
---------------------------------------





v0.10.2 - Fri, 10 Jun 2016 12:50:39 GMT
---------------------------------------

- [741c882](../../commit/741c882) [fixed] rm jsx imports



v0.10.1 - Thu, 09 Jun 2016 18:39:57 GMT
---------------------------------------





v0.10.0 - Thu, 09 Jun 2016 15:33:06 GMT
---------------------------------------





v0.9.12 - Fri, 20 May 2016 12:54:29 GMT
---------------------------------------





v0.9.11 - Fri, 15 Apr 2016 13:39:50 GMT
---------------------------------------





v0.9.10 - Fri, 15 Apr 2016 13:31:58 GMT
---------------------------------------





v0.9.9 - Fri, 15 Apr 2016 12:28:00 GMT
--------------------------------------

- [a2a49c8](../../commit/a2a49c8) [fixed] consistent handling of end dates as _exclusive_ ranges
- [1c12b16](../../commit/1c12b16) [fixed] DST issue with events
- [f526e23](../../commit/f526e23) [added] onSelecting callback
- [18c0234](../../commit/18c0234) [fixed] incorrect page offset
- [4eeacd4](../../commit/4eeacd4) [fixed] more cross browser flex issues.
- [2dc61ec](../../commit/2dc61ec) [fixed] add minHeight for week view overflow



v0.9.8 - Thu, 14 Jan 2016 19:35:07 GMT
--------------------------------------

- [5fa7012](../../commit/5fa7012) [fixed] Incorrect gutter widths



v0.9.7 - Sun, 13 Dec 2015 22:01:09 GMT
--------------------------------------

- [ebf8908](../../commit/ebf8908) [fixed] agenda header display



v0.9.6 - Sun, 13 Dec 2015 21:39:49 GMT
--------------------------------------

- [a69600b](../../commit/a69600b) [fixed] Pass correct date to DaySlot for selections.
- [5ac5c30](../../commit/5ac5c30) [fixed] reset gutter widths before calculations



v0.9.5 - Wed, 02 Dec 2015 18:09:32 GMT
--------------------------------------

- [c2e8aa4](../../commit/c2e8aa4) [fixed] accidental breaking change on the localizers
- [dc90943](../../commit/dc90943) [fixed] some style issues
- [ea8e085](../../commit/ea8e085) [added] modern globalize support
- [4b3d3ba](../../commit/4b3d3ba) [fixed] have gutter use culture prop
- [7922882](../../commit/7922882) [fixed] better gutter width detection



v0.9.4 - Sun, 29 Nov 2015 02:19:49 GMT
--------------------------------------

- [a41c9f9](../../commit/a41c9f9) [added] right-to-left support
- [8bb6589](../../commit/8bb6589) [fixed] properly select time ranges with min/max set



v0.9.3 - Sat, 28 Nov 2015 20:00:24 GMT
--------------------------------------

- [fff1914](../../commit/fff1914) [fixed] pass culture to View



v0.9.2 - Thu, 12 Nov 2015 23:34:33 GMT
--------------------------------------

- [58f008f](../../commit/58f008f) [fixed] none integer slot spaces (again)
- [f2084ef](../../commit/f2084ef) [fixed] month event rows not fitting in their cells
- [73e449d](../../commit/73e449d) [fixed] day view clicks return the correct slot



v0.9.1 - Thu, 12 Nov 2015 14:52:20 GMT
--------------------------------------

- [d5a0d20](../../commit/d5a0d20) [fixed] month event rows not fitting in their cells
- [f4b18d6](../../commit/f4b18d6) [fixed] day view clicks return the correct slot



v0.9.0 - Tue, 03 Nov 2015 11:33:38 GMT
--------------------------------------

- [ee53bbc](../../commit/ee53bbc) [changed] default "show more" behavior navigates to day view
- [80aa08f](../../commit/80aa08f) [added] popupOffset prop for configuring distance from viewport edge



v0.8.3 - Thu, 29 Oct 2015 06:07:22 GMT
--------------------------------------

- [d98af8d](../../commit/d98af8d) [added] edge detection for event popup
- [69b092d](../../commit/69b092d) [fixed] accidental hash changes
- [0351b71](../../commit/0351b71) [fixed] incorrect 'show more' layout



v0.8.2 - Wed, 28 Oct 2015 08:09:39 GMT
--------------------------------------

- [892af3d](../../commit/892af3d) [fixed] use correct handler for "show more"



v0.8.1 - Wed, 28 Oct 2015 07:58:13 GMT
--------------------------------------

- [4560ff7](../../commit/4560ff7) [changed] better thin event title collapse
- [0eeb43f](../../commit/0eeb43f) [fixed] event layout sort
- [0574eed](../../commit/0574eed) [fixed] show more row layout issues
- [7ee9959](../../commit/7ee9959) [added] title to time view events
- [c07d0ab](../../commit/c07d0ab) [changed] better event overlays for overlapping events



v0.8.0 - Mon, 26 Oct 2015 12:32:54 GMT

src/Calendar.js
import PropTypes from 'prop-types';
import React from 'react';
import uncontrollable from 'uncontrollable';
import cn from 'classnames';
import {
    accessor
  , elementType
  , dateFormat
  , dateRangeFormat
  , views as componentViews } from './utils/propTypes';

import { notify } from './utils/helpers';
import { navigate, views } from './utils/constants';
import defaultFormats from './formats';
import viewLabel from './utils/viewLabel';
import moveDate from './utils/move';
import VIEWS from './Views';
import Toolbar from './Toolbar';
import EventWrapper from './EventWrapper';
import BackgroundWrapper from './BackgroundWrapper';

import omit from 'lodash/omit';
import defaults from 'lodash/defaults';
import transform from 'lodash/transform';
import mapValues from 'lodash/mapValues';

function viewNames(_views){
  return !Array.isArray(_views) ? Object.keys(_views) : _views
}

function isValidView(view, { views: _views }) {
  let names = viewNames(_views)
  return names.indexOf(view) !== -1
}

let now = new Date();

/**
 * react-big-calendar is full featured Calendar component for managing events and dates. It uses
 * modern `flexbox` for layout making it super responsive and performant. Leaving most of the layout heavy lifting
 * to the browser. __note:__ The default styles use `height: 100%` which means your container must set an explicit
 * height (feel free to adjust the styles to suit your specific needs).
 *
 * Big Calendar is unopiniated about editing and moving events, prefering to let you implement it in a way that makes
 * the most sense to your app. It also tries not to be prescriptive about your event data structures, just tell it
 * how to find the start and end datetimes and you can pass it whatever you want.
 *
 * One thing to note is that, `react-big-calendar` treats event start/end dates as an _exclusive_ range.
 * which means that the event spans up to, but not including, the end date. In the case
 * of displaying events on whole days, end dates are rounded _up_ to the next day. So an
 * event ending on `Apr 8th 12:00:00 am` will not appear on the 8th, whereas one ending
 * on `Apr 8th 12:01:00 am` will. If you want _inclusive_ ranges consider providing a
 * function `endAccessor` that returns the end date + 1 day for those events that end at midnight.
 */
class Calendar extends React.Component {
 static propTypes = {

   /**
    * Props passed to main calendar `<div>`.
    */
   elementProps: PropTypes.object,

   /**
    * The current date value of the calendar. Determines the visible view range
    *
    * @controllable onNavigate
    */
   date: PropTypes.instanceOf(Date),

   /**
    * Shows certain dates as enabled/disabled
    *
    * ```js
    * {
    *   mon: [
    *     { start: 450, end: 600 },
    *     { start: 900, end: 1080 }
    *   ],
    *   tue: ...,
    *   '2017-06-15': [
    *     { start: 420, end: 800 }
    *   ]
    * }
    * ```
    */
   enabledHours: PropTypes.object,

   /**
    * The current view of the calendar.
    *
    * @default 'month'
    * @controllable onView
    */
   view: PropTypes.string,

   /**
    * An array of event objects to display on the calendar
    */
   events: PropTypes.arrayOf(PropTypes.object),

   /**
    * Callback fired when the `date` value changes.
    *
    * @controllable date
    */
   onNavigate: PropTypes.func,

   /**
    * Callback fired when the `view` value changes.
    *
    * @controllable date
    */
   onView: PropTypes.func,

   /**
    * A callback fired when a date selection is made. Only fires when `selectable` is `true`.
    *
    * ```js
    * (
    *   slotInfo: {
    *     start: Date,
    *     end: Date,
    *     slots: Array<Date>,
    *     action: "select" | "click"
    *   }
    * ) => any
    * ```
    */
   onSelectSlot: PropTypes.func,

   /**
    * Callback fired when a calendar event is selected.
    *
    * ```js
    * (event: Object, e: SyntheticEvent) => any
    * ```
    *
    * @controllable selected
    */
   onSelectEvent: PropTypes.func,

   /**
    * Callback fired when dragging a selection in the Time views.
    *
    * Returning `false` from the handler will prevent a selection.
    *
    * ```js
    * (range: { start: Date, end: Date }) => ?boolean
    * ```
    */
   onSelecting: PropTypes.func,
   /**
    * The selected event, if any.
    */
   selected: PropTypes.object,

   /**
    * An array of built-in view names to allow the calendar to display.
    *
    * @type Calendar.Views ('month'|'week'|'work_week'|'day'|'agenda')
    * @default ['month', 'week', 'day', 'agenda']
    */
   views: componentViews,

   /**
    * The string name of the destination view for drill-down actions, such
    * as clicking a date header, or the truncated events links. If
    * `getDrilldownView` is also specified it will be used instead.
    *
    * Set to `null` to disable drill-down actions.
    *
    * ```js
    * <BigCalendar
    *   drilldownView="agenda"
    * />
    * ```
    */
   drilldownView: PropTypes.string,

   /**
    * Functionally equivalent to `drilldownView`, but accepts a function
    * that can return a view name. It's useful for customizing the drill-down
    * actions depending on the target date and triggering view.
    *
    * Return `null` to disable drill-down actions.
    *
    * ```js
    * <BigCalendar
    *   getDrilldownView={(targetDate, currentViewName, configuredViewNames) =>
    *     if (currentViewName === 'month' && configuredViewNames.includes('week'))
    *       return 'week'
    *
    *     return null;
    *   }}
    * />
    * ```
    */
   getDrilldownView: PropTypes.func,

   /**
    * Determines whether the toolbar is displayed
    */
   toolbar: PropTypes.bool,

   /**
    * Show truncated events in an overlay when you click the "+_x_ more" link.
    */
   popup: PropTypes.bool,

   /**
    * Distance in pixels, from the edges of the viewport, the "show more" overlay should be positioned.
    *
    * ```js
    * <BigCalendar popupOffset={30}/>
    * <BigCalendar popupOffset={{x: 30, y: 20}}/>
    * ```
    */
   popupOffset: PropTypes.oneOfType([
     PropTypes.number,
     PropTypes.shape({ x: PropTypes.number, y: PropTypes.number })
   ]),
   /**
    * Allows mouse selection of ranges of dates/times.
    *
    * The 'ignoreEvents' option prevents selection code from running when a
    * drag begins over an event. Useful when you want custom event click or drag
    * logic
    */
   selectable: PropTypes.oneOf([true, false, 'ignoreEvents']),

   /**
    * Determines the selectable time increments in week and day views
    */
   step: PropTypes.number,

   /**
    * The number of slots per "section" in the time grid views. Adjust with `step`
    * to change the default of 1 hour long groups, with 30 minute slots.
    */
   timeslots: PropTypes.number,

   /**
    *Switch the calendar to a `right-to-left` read direction.
    */
   rtl: PropTypes.bool,

   /**
    * Optionally provide a function that returns an object of className or style props
    * to be applied to the the event node.
    *
    * ```js
    * (
    * 	event: Object,
    * 	start: Date,
    * 	end: Date,
    * 	isSelected: boolean
    * ) => { className?: string, style?: Object }
    * ```
    */
   eventPropGetter: PropTypes.func,

   /**
    * Optionally provide a function that returns an object of className or style props
    * to be applied to the the time-slot node. Caution! Styles that change layout or
    * position may break the calendar in unexpected ways.
    *
    * ```js
    * (
    * 	date: Date,
    * ) => { className?: string, style?: Object }
    * ```
    */
   slotPropGetter: PropTypes.func,

   /**
    * Accessor for the event title, used to display event information. Should
    * resolve to a `renderable` value.
    *
    * ```js
    * string | (event: Object) => any
    * ```
    *
    * @type {(func|string)}
    */
   titleAccessor: accessor,

   /**
    * In week and day views, will display time gutter
    */
   showTimeGutter: PropTypes.bool,

   /**
    * Will not render the header above timegrid and month view
    */
   showHeader: PropTypes.bool,

   /**
    * Determines whether the event should be considered an "all day" event and ignore time.
    * Must resolve to a `boolean` value.
    *
    * ```js
    * string | (event: Object) => boolean
    * ```
    *
    * @type {(func|string)}
    */
   allDayAccessor: accessor,

   /**
    * The start date/time of the event. Must resolve to a JavaScript `Date` object.
    *
    * ```js
    * string | (event: Object) => Date
    * ```
    *
    * @type {(func|string)}
    */
   startAccessor: accessor,

   /**
    * The end date/time of the event. Must resolve to a JavaScript `Date` object.
    *
    * ```js
    * string | (event: Object) => Date
    * ```
    *
    * @type {(func|string)}
    */
   endAccessor: accessor,

   /**
    * Constrains the minimum _time_ of the Day and Week views.
    */
   min: PropTypes.instanceOf(Date),

   /**
    * Constrains the maximum _time_ of the Day and Week views.
    */
   max: PropTypes.instanceOf(Date),

   /**
    * Determines how far down the scroll pane is initially scrolled down.
    */
   scrollToTime: PropTypes.instanceOf(Date),

   /**
    * Specify a specific culture code for the Calendar.
    *
    * **Note: it's generally better to handle this globally via your i18n library.**
    */
   culture: PropTypes.string,

   /**
    * Localizer specific formats, tell the Calendar how to format and display dates.
    *
    * `format` types are dependent on the configured localizer; both Moment and Globalize
    * accept strings of tokens according to their own specification, such as: `'DD mm yyyy'`.
    *
    * ```jsx
    * let formats = {
    *   dateFormat: 'dd',
    *
    *   dayFormat: (date, culture, localizer) =>
    *     localizer.format(date, 'DDD', culture),
    *
    *   dayRangeHeaderFormat: ({ start, end }, culture, local) =>
    *     local.format(start, { date: 'short' }, culture) + ' — ' +
    *     local.format(end, { date: 'short' }, culture)
    * }
    *
    * <Calendar formats={formats} />
    * ```
    *
    * All localizers accept a function of
    * the form `(date: Date, culture: ?string, localizer: Localizer) -> string`
    */
   formats: PropTypes.shape({
     /**
      * Format for the day of the month heading in the Month view.
      * e.g. "01", "02", "03", etc
      */
     dateFormat,

     /**
      * A day of the week format for Week and Day headings,
      * e.g. "Wed 01/04"
      *
      */
     dayFormat: dateFormat,

     /**
      * Week day name format for the Month week day headings,
      * e.g: "Sun", "Mon", "Tue", etc
      *
      */
     weekdayFormat: dateFormat,

     /**
      * The timestamp cell formats in Week and Time views, e.g. "4:00 AM"
      */
     timeGutterFormat: dateFormat,

     /**
      * Toolbar header format for the Month view, e.g "2015 April"
      *
      */
     monthHeaderFormat: dateFormat,

     /**
      * Toolbar header format for the Week views, e.g. "Mar 29 - Apr 04"
      */
     dayRangeHeaderFormat: dateRangeFormat,

     /**
      * Toolbar header format for the Day view, e.g. "Wednesday Apr 01"
      */
     dayHeaderFormat: dateFormat,

     /**
      * Toolbar header format for the Agenda view, e.g. "4/1/2015 — 5/1/2015"
      */
     agendaHeaderFormat: dateFormat,

     /**
      * A time range format for selecting time slots, e.g "8:00am — 2:00pm"
      */
     selectRangeFormat: dateRangeFormat,

     agendaDateFormat: dateFormat,
     agendaTimeFormat: dateFormat,
     agendaTimeRangeFormat: dateRangeFormat,

     /**
      * Time range displayed on events.
      */
     eventTimeRangeFormat: dateRangeFormat
   }),

   /**
    * Customize how different sections of the calendar render by providing custom Components.
    * In particular the `Event` component can be specified for the entire calendar, or you can
    * provide an individual component for each view type.
    *
    * ```jsx
    * let components = {
    *   event: MyEvent, // used by each view (Month, Day, Week)
    *   toolbar: MyToolbar,
    *   agenda: {
    *   	 event: MyAgendaEvent // with the agenda view use a different component to render events
    *   }
    * }
    * <Calendar components={components} />
    * ```
    */
   components: PropTypes.shape({
     event: elementType,
     eventWrapper: elementType,
     dayWrapper: elementType,
     dateCellWrapper: elementType,

     toolbar: elementType,

     agenda: PropTypes.shape({
       date: elementType,
       time: elementType,
       event: elementType
     }),

     day: PropTypes.shape({
       header: elementType,
       event: elementType
     }),
     week: PropTypes.shape({
       header: elementType,
       event: elementType
     }),
     month: PropTypes.shape({
       header: elementType,
       dateHeader: elementType,
       event: elementType
     })
   }),

   /**
    * String messages used throughout the component, override to provide localizations
    */
   messages: PropTypes.shape({
     allDay: PropTypes.node,
     previous: PropTypes.node,
     next: PropTypes.node,
     today: PropTypes.node,
     month: PropTypes.node,
     week: PropTypes.node,
     day: PropTypes.node,
     agenda: PropTypes.node,
     showMore: PropTypes.func
   })
 };

 static defaultProps = {
   elementProps: {},
   popup: false,
   toolbar: true,
   view: views.MONTH,
   views: [views.MONTH, views.WEEK, views.DAY, views.AGENDA],
   date: now,
   step: 30,

   drilldownView: views.DAY,

   titleAccessor: 'title',
   allDayAccessor: 'allDay',
   startAccessor: 'start',
   endAccessor: 'end',

   showTimeGutter: true,
   showHeader: true,
 };

 getViews = () => {
   const views = this.props.views;

   if (Array.isArray(views)) {
     return transform(views, (obj, name) => obj[name] = VIEWS[name], {});
   }

   if (typeof views === 'object') {
     return mapValues(views, (value, key) => {
       if (value === true) {
         return VIEWS[key];
       }

       return value;
     });
   }

   return VIEWS;
 };

 getView = () => {
   const views = this.getViews();

   return views[this.props.view];
 };

 getDrilldownView = (date) => {
   const { view, drilldownView, getDrilldownView } = this.props

   if (!getDrilldownView) return drilldownView

   return getDrilldownView(date, view, Object.keys(this.getViews()));
 };

 render() {
   let {
       view, toolbar, events
     , enabledHours
     , culture
     , components = {}
     , formats = {}
     , style
     , className
     , elementProps
     , date: current
     , ...props } = this.props;

   formats = defaultFormats(formats)

   let View = this.getView();
   let names = viewNames(this.props.views)

   let viewComponents = defaults(
     components[view] || {},
     omit(components, names),
     {
       eventWrapper: EventWrapper,
       dayWrapper: BackgroundWrapper,
       dateCellWrapper: BackgroundWrapper
     }
   )

   let ToolbarToRender = components.toolbar || Toolbar

   return (
     <div
       {...elementProps}
       className={cn('rbc-calendar', className, {
         'rbc-rtl': props.rtl
       })}
       style={style}
     >
       {toolbar &&
         <ToolbarToRender
           date={current}
           view={view}
           views={names}
           label={viewLabel(current, view, formats, culture)}
           onViewChange={this.handleViewChange}
           onNavigate={this.handleNavigate}
           messages={this.props.messages}
         />
       }
       <View
         ref='view'
         {...props}
         {...formats}
         enabledHours={enabledHours}
         culture={culture}
         formats={undefined}
         events={events}
         date={current}
         components={viewComponents}
         getDrilldownView={this.getDrilldownView}
         onNavigate={this.handleNavigate}
         onDrillDown={this.handleDrillDown}
         onSelectEvent={this.handleSelectEvent}
         onSelectSlot={this.handleSelectSlot}
         onShowMore={this._showMore}
       />
     </div>
   );
 }

 handleNavigate = (action, newDate) => {
   let { view, date, onNavigate } = this.props;
   let ViewComponent = this.getView();

   date = moveDate(action, newDate || date, ViewComponent)

   onNavigate(date, view, action)
 };

 handleViewChange = (view) => {
   if (view !== this.props.view && isValidView(view, this.props))
     this.props.onView(view)
 };

 handleSelectEvent = (...args) => {
   notify(this.props.onSelectEvent, args)
 };

 handleSelectSlot = (slotInfo) => {
   notify(this.props.onSelectSlot, slotInfo)
 };

 handleDrillDown = (date, view) => {
   if (view)
     this.handleViewChange(view)

   this.handleNavigate(navigate.DATE, date)
 };
}

export default uncontrollable(Calendar, {
  view: 'onView',
  date: 'onNavigate',
  selected: 'onSelectEvent'
})
